src/Microsoft.ComponentDetection.Detectors/conda/CondaLockComponentDetector.cs

```diff
@@ -0,0 +1,126 @@
+namespace Microsoft.ComponentDetection.Detectors.Poetry;
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.ComponentDetection.Contracts;
+using Microsoft.ComponentDetection.Contracts.Internal;
+using Microsoft.ComponentDetection.Contracts.TypedComponent;
+using Microsoft.ComponentDetection.Detectors.Poetry.Contracts;
+using Microsoft.Extensions.Logging;
+using YamlDotNet.Serialization;
+
+public class CondaLockComponentDetector : FileComponentDetector, IDefaultOffComponentDetector
+{
+    public CondaLockComponentDetector(
+        IComponentStreamEnumerableFactory componentStreamEnumerableFactory,
+        IObservableDirectoryWalkerFactory walkerFactory,
+        ILogger<CondaLockComponentDetector> logger)
+    {
+        this.ComponentStreamEnumerableFactory = componentStreamEnumerableFactory;
+        this.Scanner = walkerFactory;
+        this.Logger = logger;
+    }
+
+    public override string Id => "CondaLock";
+
+    public override IList<string> SearchPatterns { get; } = new List<string> { "conda-lock.yml", "*.conda-lock.yml" };
+
+    public override IEnumerable<ComponentType> SupportedComponentTypes => new[] { ComponentType.Conda, ComponentType.Pip };
+
+    public override int Version { get; } = 1;
+
+    public override IEnumerable<string> Categories => new List<string> { "Python" };
+
+    /// <inheritdoc/>
+    protected override Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+    {
+        var singleFileComponentRecorder = processRequest.SingleFileComponentRecorder;
+
+        this.Logger.LogDebug("Found cond-lock file: {YamlFile}", processRequest.ComponentStream.Location);
+        try
+        {
+            // Parse conda lock file
+            var condaLock = this.ParseCondaLock(processRequest);
+
+            // Parse conda environments to get explicit dependencies
+            var explicitDependencies = new List<string>();
+            if (condaLock != null && condaLock.Metadata != null && condaLock.Metadata.Sources != null)
+            {
+                foreach (var environment in condaLock.Metadata.Sources)
+                {
+                    var environmentFilePath = Path.Combine(processRequest.ComponentStream.Location, "..", environment);
+                    explicitDependencies.AddRange(this.ParseExplicitDependencies(environmentFilePath));
```

Comment:
Is the environment file necessary to capture the dependencies? Shouldn't those be declared in conda-lock file directly or why the need to open a separate file? I don't have much context on this manifest, so it may be necessary I just want to understand the reasoning behind that behavior.

Reply:
All dependencies are in the conda-lock file. The initial idea here was to use the environment.yml file to identify the packages that have been referenced directly to set the isReferencedDirectly flag correctly. 

I fully agree that parsing another file complicates the code though. I refactored the code to identify directly referenced packages by searching for packages that don't have any ancestors. This way we can avoid parsing the additional file.